File: problems/1306. Jump Game III/main.go
func canReach(arr []int, start int) bool {
	visited := make([]bool, len(arr))

	var helper func(start int) bool
	helper = func(start int) bool {
		if start >= len(arr) || start < 0 || visited[start] {
			return false
		}
		if arr[start] == 0 {
			return true
		}

		visited[start] = true

		return helper(start+arr[start]) || helper(start-arr[start])
	}

	return helper(start)
}

func canReach(arr []int, start int) bool {
	n := len(arr)
	visited := make([]bool, n)
	queue := []int{start}
	visited[start] = true

	for len(queue) > 0 {
		i := queue[0]
		queue = queue[1:]

		if arr[i] == 0 {
			return true
		}

		for _, next_i := range []int{i + arr[i], i - arr[i]} {
			if next_i >= 0 && next_i < n && !visited[next_i] {
				visited[next_i] = true
				queue = append(queue, next_i)
			}
		}
	}

	return false
}

File: problems/1306. Jump Game III/main.py
class Solution(object):
    def canReach(self, arr, start):
        """
        :type arr: List[int]
        :type start: int
        :rtype: bool
        """
        visited = [False] * len(arr)

        def helper(start):
            if start >= len(arr) or start < 0 or visited[start]:
                return False
            if arr[start] == 0:
                return True

            visited[start] = True

            return helper(start + arr[start]) or helper(start - arr[start])

        return helper(start)


class Solution(object):
    def canReach(self, arr, start):
        """
        :type arr: List[int]
        :type start: int
        :rtype: bool
        """
        n = len(arr)
        visited = [False] * n
        queue = deque([start])
        visited[start] = True

        while queue:
            i = queue.popleft()

            if arr[i] == 0:
                return True

            for next_i in [i + arr[i], i - arr[i]]:
                if 0 <= next_i < n and not visited[next_i]:
                    visited[next_i] = True
                    queue.append(next_i)

        return False
